src/rmt.rs

```diff
@@ -0,0 +1,332 @@
+        self
+    }
+
+    pub fn carrier_freq_hz(mut self, freq: u32) -> Self {
+        self.config.__bindgen_anon_1.tx_config.carrier_freq_hz = freq;
+        self
+    }
+
+    // TODO: Restrict 0-100 using a newtype.
+    pub fn carrier_duty_percent(mut self, percent: u8) -> Self {
+        self.config.__bindgen_anon_1.tx_config.carrier_duty_percent = percent;
+        self
+    }
+
+    pub fn carrier_level(mut self, level: PinState) -> Self {
+        self.config.__bindgen_anon_1.tx_config.carrier_level = level as u32;
+        self
+    }
+
+    pub fn idle_level(mut self, level: PinState) -> Self {
+        self.config.__bindgen_anon_1.tx_config.idle_level = level as u32;
+        self
+    }
+
+    pub fn idle_output_enable(mut self, enabled: bool) -> Self {
+        self.config.__bindgen_anon_1.tx_config.idle_output_en = enabled;
+        self
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct PulseTicks(u16);
+
+impl PulseTicks {
+    const MAX: u16 = 32767;
+
+    /// PulseTick needs to be unsigned 15 bits: 0-32767 inclusive.
+    ///
+    /// It will return `None` if it's out of range.
+    pub fn new(v: u16) -> Option<Self> {
+        if v > 32767 {
+            // TODO: Not sure of a nicer way to deal with this? Allow an overflow like rust does?
+            None
+        } else {
+            Some(Self(v))
+        }
+    }
+
+    /// Use the maximum value of 32767.
+    pub fn max() -> Self {
+        Self(Self::MAX)
+    }
+}
+
+pub struct Writer {
+    config: rmt_config_t,
+
+    // This must be ManuallyDrop to ensure that it isn't automatically dropped before the driver is
+    // done using the items.
+    // TODO: Manually drop this!
+    items: ManuallyDrop<Vec<rmt_item32_t>>,
+
+    // An item that has had only its first half populated.
+    half_inserted: Option<rmt_item32_t>,
+
+    ticks_per_ns: Option<f32>,
+}
+
+impl Writer {
+    pub fn new(config: WriterConfig) -> Result<Writer, EspError> {
+        let s = Writer {
+            config: config.config,
+            items: Default::default(),
+            half_inserted: None,
+            ticks_per_ns: None,
+        };
+
+        unsafe {
+            esp!(rmt_config(&s.config))?;
+            esp!(rmt_driver_install(s.config.channel as u32, 0, 0))?;
+        }
+
+        Ok(s)
+    }
+
+    pub fn counter_clock(&self) -> Result<u32, EspError> {
+        let mut ticks_hz: u32 = 0;
+        esp!(unsafe { rmt_get_counter_clock(self.config.channel, &mut ticks_hz) })?;
+        Ok(ticks_hz)
+    }
+
+    /// Creates a Pulse based on Nanoseconds.
+    ///
+    /// Returns None if the resulting ticks from the ns conversion is too high.
+    /// See `PulseTicks` for details.
+    ///
+    /// This function exists on `Writer` because of the internal call to `rmt_get_counter_clock`
+    /// requires a channel which is managed by `Writer`.
+    pub fn pulse_ns(&mut self, pin_state: PinState, ns: u32) -> Result<Option<Pulse>, EspError> {
```

Comment:
> Using FP operations (f32) is not ideal
I did consider that but couldn't see a way with smaller sized units (I'm worried about performance). Will investigate options, maybe using a larger unit assuming Rust+ESP can handle 64 bit numbers.

> core::time::Duration
Same reasoning--I was worried about performance. I am happy to change if it's not a concern.

> Perhaps better to make this a method of Pulse itself, similar to the existing new, but taking ticks_hz as a parameter?
I did have it like this before, but I didn't want the user to have to manually call `rmt_get_counter_clock`/`fn counter_clock()` and track `ticks_hz`, only because of convenience to the user.

Reply:
Re FP and handling 64 bit numbers:

Rust+ESP can handle everything Rust can handle. u32, u64, u128 - you name it. And FP. Except that FP is the slowest of all, especially on hardware that does not have native FP support (as in riscv32imc, which is esp32c3). FP also affects FreeRTOS context switching and specifically for ESP-IDF has "[interesting](https://docs.espressif.com/projects/esp-idf/en/v4.2.2/esp32/api-guides/freertos-smp.html#floating-point-arithmetic)" (as in unwanted) consequences.

As a background information, even on big hardware, using FP in driver code (Linux, Windows) is a big "no, no".

So please let's remove that if possible.

Re `core::time:Duration`: it is just a `u64` + `u32`, so performance concerns might be exaggerated.

Reply:
Thanks for the info on how bad FP is. I'm not used to working with embedded so I didn't realise how bad it is.

Duration implementation is done and no floats in the code now. Also part of it was move the Duration -> ticks calculation into Pulse/PulseDuration, outside of Writer. It's slightly more verbose for the user, but at the same time the code is less complex in Writer, so I think that's a win.